walrus-attributes-indexer/src/handlers/walrus_blob.rs

```diff
@@ -0,0 +1,471 @@
+                .do_update()
+                .set((
+                    walrus_blob::deleted.eq(true),
+                    walrus_blob::cp_sequence_number.eq(excluded(walrus_blob::cp_sequence_number)),
+                ))
+                .filter(
+                    walrus_blob::cp_sequence_number.lt(excluded(walrus_blob::cp_sequence_number)),
+                )
+                .execute(conn)
+                .await?;
+
+            let historical_deletes: Vec<StoredWalrusBlobHistorical> =
+                deletes.iter().map(|v| v.into()).collect();
+
+            total_affected += diesel::insert_into(walrus_blob_historical::table)
+                .values(historical_deletes)
+                .on_conflict_do_nothing()
+                .execute(conn)
+                .await?;
+        }
+
+        Ok(total_affected)
+    }
+}
+
+impl FieldCount for ProcessedWalrusMetadata {
+    const FIELD_COUNT: usize = StoredWalrusBlob::FIELD_COUNT;
+}
+
+impl TryInto<StoredWalrusBlob> for &ProcessedWalrusMetadata {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> Result<StoredWalrusBlob> {
+        let blob_object: Blob = bcs::from_bytes(
+            self.parent_object
+                .data
+                .try_as_move()
+                .ok_or_else(|| anyhow::anyhow!("Parent object is not a Move object"))?
+                .contents(),
+        )?;
+
+        let address_owner = match self.parent_object.owner() {
+            Owner::AddressOwner(id) => id.to_vec(),
+            _ => {
+                return Err(anyhow::anyhow!(
+                    "Parent object's owner is not an address owner"
+                ))
+            }
+        };
+
+        Ok(StoredWalrusBlob {
+            dynamic_field_id: self.dynamic_field_id.to_vec(),
+            cp_sequence_number: self.cp_sequence_number as i64,
+            file_path: self.file_path.clone(),
+            owner_id: self.parent_object.id().to_vec(),
+            address_owner,
+            blob_id: blob_object.blob_id.0.to_vec(),
+            deleted: self.deleted,
+        })
+    }
+}
+
+impl From<&StoredWalrusBlob> for StoredWalrusBlobHistorical {
+    fn from(value: &StoredWalrusBlob) -> Self {
+        if value.deleted {
+            return StoredWalrusBlobHistorical {
+                dynamic_field_id: value.dynamic_field_id.clone(),
+                cp_sequence_number: value.cp_sequence_number,
+                owner_id: None,
+                address_owner: None,
+                file_path: None,
+                blob_id: None,
+            };
+        }
+        StoredWalrusBlobHistorical {
+            dynamic_field_id: value.dynamic_field_id.clone(),
+            cp_sequence_number: value.cp_sequence_number,
+            owner_id: Some(value.owner_id.clone()),
+            address_owner: Some(value.address_owner.clone()),
+            file_path: Some(value.file_path.clone()),
+            blob_id: Some(value.blob_id.clone()),
+        }
```

Comment:
Not sure how I feel about using `From` for this -- it feels borderline.

```suggestion
        if value.deleted {
            StoredWalrusBlobHistorical {
                dynamic_field_id: value.dynamic_field_id.clone(),
                cp_sequence_number: value.cp_sequence_number,
                owner_id: None,
                address_owner: None,
                file_path: None,
                blob_id: None,
            }
        } else {
            StoredWalrusBlobHistorical {
                dynamic_field_id: value.dynamic_field_id.clone(),
                cp_sequence_number: value.cp_sequence_number,
                owner_id: Some(value.owner_id.clone()),
                address_owner: Some(value.address_owner.clone()),
                file_path: Some(value.file_path.clone()),
                blob_id: Some(value.blob_id.clone()),
            }
        }
```

Reply:
i can change this to a regular method on StoredWalrusBlob